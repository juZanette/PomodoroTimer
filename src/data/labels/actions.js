import fs from '../../firebase/firebase'

export const setFullscreenDialog = (fullscreenDialog) => ({
  type: 'SET_FULLSCREEN_DIALOG',
  fullscreenDialog,
})

export const setDesktopDialog = (desktopDialog) => ({
  type: 'SET_DESKTOP_DIALOG',
  desktopDialog,
})

export const setDeleteAlert = (deleteAlert) => ({
  type: 'SET_DELETE_ALERT',
  deleteAlert,
})

export const setMenuOpened = (menuOpened) => ({
  type: 'SET_MENU_OPENED',
  menuOpened,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})

export const setLabelEditting = (labelEditting) => ({
  type: 'SET_LABEL_EDITTING',
  labelEditting,
})

export const setLabelSelected = (labelSelected) => ({
  type: 'SET_LABEL_SELECTED',
  labelSelected,
})

export const addLabel = (label) => ({
  type: 'ADD_LABEL',
  label,
})

export const startAddLabel = (label) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const ref = await fs.collection(`users/${uid}/labels`).add(label)

    dispatch(
      addLabel({
        id: ref.id,
        ...label,
      })
    )

    return ref
  }
}

export const editLabel = (id, updates) => ({
  type: 'EDIT_LABEL',
  id,
  updates,
})

export const startEditLabel = (id, updates) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    await fs.doc(`users/${uid}/labels/${id}`).update({
      ...updates,
    })

    dispatch(editLabel(id, updates))
  }
}

export const deleteLabel = (id) => ({
  type: 'DELETE_LABEL',
  id,
})

export const startDeleteLabel = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid
    await fs.doc(`users/${uid}/labels/${id}`).delete()

    dispatch(deleteLabel(id))
  }
}

export const setLabels = (labels) => ({
  type: 'SET_LABELS',
  labels,
})

export const startSetLabels = () => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid

    const labelsRef = await fs.collection(`users/${uid}/labels`).get()

    const labels = labelsRef.docs.map((label) => ({
      id: label.id,
      ...label.data(),
    }))

    dispatch(setLabels(labels))

    return labelsRef
  }
}
