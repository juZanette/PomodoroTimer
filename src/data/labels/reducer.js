const initialState = {
  dialogOpened: false,
  labelToEdit: null,
  formValue: {
    name: null,
    color: null,
  },
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
