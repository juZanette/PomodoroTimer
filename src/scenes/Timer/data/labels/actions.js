export const setDialogOpened = (dialogOpened) => ({
  type: 'SET_DIALOG_OPENED',
  dialogOpened,
})

export const setAlertOpened = (alertOpened) => ({
  type: 'SET_ALERT_OPENED',
  alertOpened,
})

export const setLabelToEdit = (labelToEdit) => ({
  type: 'SET_LABEL_TO_EDIT',
  labelToEdit,
})

export const setFormValue = (formValue) => ({
  type: 'SET_FORM_VALUE',
  formValue,
})
