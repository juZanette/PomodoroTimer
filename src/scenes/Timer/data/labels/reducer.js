const initialState = {
  dialogOpened: false,
  formDialogOpened: false,
  deleteAlert: {
    opened: false,
    labelToDelete: null,
  },
  labelToEdit: null,
  formValue: null,
}

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_DIALOG_OPENED':
      return {
        ...state,
        dialogOpened: action.dialogOpened,
      }
    case 'SET_FORM_DIALOG':
      return {
        ...state,
        formDialogOpened: action.formDialogOpened,
      }
    case 'SET_DELETE_ALERT':
      const { opened, labelToDelete } = action.deleteAlert

      return {
        ...state,
        deleteAlert: {
          opened,
          labelToDelete,
        },
      }
    case 'SET_LABEL_TO_EDIT':
      return {
        ...state,
        labelToEdit: action.labelToEdit,
      }
    case 'SET_FORM_VALUE':
      return {
        ...state,
        formValue: action.formValue,
      }
    default:
      return state
  }
}
