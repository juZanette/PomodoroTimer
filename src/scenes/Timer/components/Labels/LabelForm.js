import React, { useState } from 'react'
import TextField from '@material-ui/core/TextField'
import MenuItem from '@material-ui/core/MenuItem'
import styled from 'styled-components'

import red from '@material-ui/core/colors/red'
import deepPurple from '@material-ui/core/colors/deepPurple'
import blue from '@material-ui/core/colors/blue'
import cyan from '@material-ui/core/colors/cyan'
import teal from '@material-ui/core/colors/teal'
import green from '@material-ui/core/colors/green'
import yellow from '@material-ui/core/colors/yellow'
import deepOrange from '@material-ui/core/colors/deepOrange'
import { useDispatch, useSelector } from 'react-redux'
import { setFormValue } from '../../data/labels/actions'

const COLORS = [
  { name: 'Red', hex: red[500], id: '111' },
  { name: 'Deep Purple', hex: deepPurple[500], id: '222' },
  { name: 'Blue', hex: blue[500], id: '333' },
  { name: 'Cyan', hex: cyan[500], id: '444' },
  { name: 'Teal', hex: teal[500], id: '555' },
  { name: 'Deep Orange', hex: deepOrange[500], id: '666' },
  { name: 'Green', hex: green[500], id: '777' },
  { name: 'Yellow', hex: yellow[500], id: '888' },
]

const ColorTextField = styled(TextField)`
  #label-color {
    display: flex;
    align-items: center;
  }
`

const ColorIndicator = styled.span`
  background: ${({ color }) => color};
  display: inline-block;
  width: 20px;
  height: 20px;
  border-radius: 50%;
  margin-right: 20px;
`

export const LabelForm = () => {
  const { labelEditting } = useSelector((state) => state.labels)

  const [name, setName] = useState(labelEditting ? labelEditting.name : '')
  const [nameValid, setNameValid] = useState(true)
  const nameHelperText = 'Fill in label name'

  const [color, setColor] = useState(labelEditting ? labelEditting.color : '')

  const dispatch = useDispatch()

  const onNameChange = (e) => {
    const value = e.target.value

    setName(value)
    dispatch(setFormValue({ name: value, color }))

    if (!!value !== nameValid) {
      setNameValid(!nameValid)
    }
  }

  const onColorChange = (e) => {
    const value = e.target.value

    setColor(value)
    dispatch(setFormValue({ name, color: value }))
  }

  const onNameBlur = () => {
    if (!name) {
      setNameValid(false)
    }
  }

  return (
    <form noValidate autoComplete="off">
      <TextField
        error={!nameValid}
        helperText={!nameValid ? nameHelperText : ' '}
        required
        id="label-name"
        label="Label name"
        variant="filled"
        name="name"
        margin="normal"
        value={name}
        onChange={onNameChange}
        onBlur={onNameBlur}
        fullWidth
        autoFocus
      />

      <ColorTextField
        select
        required
        id="label-color"
        label="Label color"
        variant="filled"
        name="color"
        margin="normal"
        value={color}
        onChange={onColorChange}
        fullWidth
      >
        {COLORS.map((color) => (
          <MenuItem key={color.id} value={color.hex}>
            <ColorIndicator color={color.hex} />
            <span>{color.name}</span>
          </MenuItem>
        ))}
      </ColorTextField>
    </form>
  )
}
