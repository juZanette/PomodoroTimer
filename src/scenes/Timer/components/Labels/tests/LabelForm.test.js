import React from 'react'
import { act } from 'react-dom/test-utils'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { createMount } from '@material-ui/core/test-utils'
import TextField from '@material-ui/core/TextField'
import labels from '../../../../../data/labels/tests/mock-data/labels'
import { LabelForm, ColorTextField } from '../LabelForm'

describe('<LabelForm />', () => {
  const mockStore = configureMockStore([thunk])

  let store
  let mount
  let wrapper

  beforeEach(() => {
    mount = createMount()

    wrapper = (labelEditting = null) => {
      const storeData = {
        labels: { labelEditting },
      }

      store = mockStore(storeData)
      store.dispatch = jest.fn()

      return mount(
        <Provider store={store}>
          <LabelForm />
        </Provider>
      )
    }
  })

  afterEach(() => {
    mount.cleanUp()
  })

  test('should render <LabelForm /> correctly', () => {
    expect(wrapper()).toMatchSnapshot()
  })

  test('should display error message if name field is invalid', () => {
    const wrapperRendered = wrapper(labels[0])

    wrapperRendered
      .find('input')
      .at(0)
      .simulate('change', { target: { value: '' } })

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(wrapperRendered.find(TextField).at(0).prop('error')).toBe(true)
  })

  test('should not display error message if name field is valid', () => {
    const wrapperRendered = wrapper(labels[0])

    wrapperRendered
      .find('input')
      .at(0)
      .simulate('change', { target: { value: `${labels[0].name} ` } })

    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(wrapperRendered.find(TextField).at(0).prop('error')).toBe(false)
  })

  test('should display error message on blur if name field is empty', () => {
    const wrapperRendered = wrapper()

    wrapperRendered
      .find('input')
      .at(0)
      .simulate('blur', { target: { value: '' } })

    expect(wrapperRendered.find(TextField).at(0).prop('error')).toBe(true)
  })

  test('should call dispatch on color value change', () => {
    const wrapperRendered = wrapper(labels[0])

    act(() => {
      wrapperRendered.find(ColorTextField).prop('onChange')({
        target: { value: '#ffeb3b' },
      })
    })

    expect(store.dispatch).toHaveBeenCalledTimes(1)
  })
})
