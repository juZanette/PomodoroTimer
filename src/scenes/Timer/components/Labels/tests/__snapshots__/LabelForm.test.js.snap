// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelForm /> should render <LabelForm /> correctly 1`] = `
<styled.form
  autoComplete="off"
  color="#7986cb"
  darkMode={0}
  noValidate={true}
>
  <WithStyles(ForwardRef(TextField))
    autoFocus={true}
    error={false}
    fullWidth={true}
    helperText=" "
    id="label-name"
    label="Label name"
    margin="normal"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    required={true}
    value=""
    variant="filled"
  />
  <Styled(WithStyles(ForwardRef(TextField)))
    fullWidth={true}
    id="label-color"
    label="Label color"
    margin="normal"
    name="color"
    onChange={[Function]}
    required={true}
    select={true}
    value=""
    variant="filled"
  >
    <WithStyles(ForwardRef(MenuItem))
      key="1"
      value="#f44336"
    >
      <styled.span
        color="#f44336"
      />
      <span>
        Red
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="2"
      value="#e91e63"
    >
      <styled.span
        color="#e91e63"
      />
      <span>
        Pink
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="3"
      value="#9c27b0"
    >
      <styled.span
        color="#9c27b0"
      />
      <span>
        Purple
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="4"
      value="#673ab7"
    >
      <styled.span
        color="#673ab7"
      />
      <span>
        Deep Purple
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="5"
      value="#3f51b5"
    >
      <styled.span
        color="#3f51b5"
      />
      <span>
        Indigo
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="6"
      value="#2196f3"
    >
      <styled.span
        color="#2196f3"
      />
      <span>
        Blue
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="7"
      value="#03a9f4"
    >
      <styled.span
        color="#03a9f4"
      />
      <span>
        Light Blue
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="8"
      value="#00bcd4"
    >
      <styled.span
        color="#00bcd4"
      />
      <span>
        Cyan
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="9"
      value="#009688"
    >
      <styled.span
        color="#009688"
      />
      <span>
        Teal
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="10"
      value="#4caf50"
    >
      <styled.span
        color="#4caf50"
      />
      <span>
        Green
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="11"
      value="#8bc34a"
    >
      <styled.span
        color="#8bc34a"
      />
      <span>
        Light Green
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="12"
      value="#cddc39"
    >
      <styled.span
        color="#cddc39"
      />
      <span>
        Lime
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="13"
      value="#ffeb3b"
    >
      <styled.span
        color="#ffeb3b"
      />
      <span>
        Yellow
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="14"
      value="#ffc107"
    >
      <styled.span
        color="#ffc107"
      />
      <span>
        Amber
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="15"
      value="#ff9800"
    >
      <styled.span
        color="#ff9800"
      />
      <span>
        Orange
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="16"
      value="#ff5722"
    >
      <styled.span
        color="#ff5722"
      />
      <span>
        Deep Orange
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="17"
      value="#795548"
    >
      <styled.span
        color="#795548"
      />
      <span>
        Brown
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="18"
      value="#9e9e9e"
    >
      <styled.span
        color="#9e9e9e"
      />
      <span>
        Grey
      </span>
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="19"
      value="#607d8b"
    >
      <styled.span
        color="#607d8b"
      />
      <span>
        Blue Grey
      </span>
    </WithStyles(ForwardRef(MenuItem))>
  </Styled(WithStyles(ForwardRef(TextField)))>
</styled.form>
`;
